stores/store.go

```diff
@@ -230,7 +230,7 @@ type MsgStore interface {
 	State() (numMessages int, byteSize uint64, err error)
 
 	// Store stores a message and returns the message sequence.
-	Store(data []byte) (uint64, error)
+	Store(msg *pb.MsgProto) (uint64, error)
```

Comment:
Are you passing a `pb.MsgProto` because you need, or because it would break filestore otherwise?

If we are to change the Store() here, I wonder if we should not change it to `seq, []byte`, and Store
would store a bag of bytes with given sequence.

On Lookup() it would return bag of bytes for this sequence (again, filestore would need fixing and 
probably not be backward compatible).

Basically, it would remove MsgProto knowledge to the Store implementation.
The server would call Store() with the serialized version of MsgProto. This would allow, on Lookup,
to send the byte array directly to subs without the need to marshal it. We would have to do unmarshal/
marshal on redelivery since we need to set the Redelivered flag in the MsgProto though.

Reply:
The reason for this change is the leader needs to assign metadata for a message (sequence and timestamp) before replicating. I'm open to suggestions here.

I like the idea of avoiding unnecessary marshals/unmarshals where possible. My profiling showed that has a pretty major impact on performance.

Reply:
(some notes to help with making a decision)

If we were to change, I realize that we would have to pass sequence and timestamp (along with byte array) to the `Store()` call since the store would still need those 2 informations (for message expiration, and lookup), which in your branch it gets from the `MsgProto` passed as a parameter.

If `FileMsgStore` were to store the given serialized `MsgProto` but without knowledge that it is a `MsgProto`, newer server would not be able to recover datastore that were created with the `0.2.2 MVP` release (for which index files did not exist). Users would have to start it with say a `0.5.0` which would create the index files on recovery, before being able to use the newer version that would not be able to recover data files without index files. (since newer server would not know that content is a `MsgProto` from which it can get sequence and timestamp).
The `0.2.2 MVP` being so old, I don't think anybody is running this version, so that would be less of an issue.

`MemStore` would have to change too, since it has a map of `MsgProto` keyed by sequence. For expiration it uses the timestamp from the `MsgProto` values. So we would need to have a struct that contains timestamp and byte array. (it kinds of duplicate the storage of timestamp since in reality it would be in the serialized data `MemoryMsgStore` is storing, but..).